File: iosApp/iosApp/Providers/RevenueCatBillingProvider.swift
import Foundation
import RevenueCat
import FirebaseAuth
import shared

class RevenueCatBillingProvider: BillingProvider {
    
    func configure(apiKey: String, userId: String?) async throws {
        Purchases.logLevel = .debug
        Purchases.configure(withAPIKey: apiKey, appUserID: userId)
    }
    
    func logIn(userId: String, email: String?) async throws {
        let _ = try await Purchases.shared.logIn(userId)
        if email != nil {
            Purchases.shared.attribution.setEmail(email)
        }
    }
    
    func setEmail(email: String) async throws {
        Purchases.shared.attribution.setEmail(email)
    }
    
    func logOut() async throws {
        if !Purchases.shared.isAnonymous {
            let _ = try await Purchases.shared.logOut()
        }
    }
    
    func getCustomerBillingInfo() async throws -> CustomerBillingInfo {
        
        let rcCustomerInfo = try await Purchases.shared.customerInfo()
        return CustomerBillingInfo(
            entitlements: rcCustomerInfo.entitlements.active.keys.map{ $0.description },
            purchases: Array(rcCustomerInfo.allPurchasedProductIdentifiers),
            managementUrl: rcCustomerInfo.managementURL?.absoluteString
        )
        
    }
    
    func getProducts() async throws -> [Product] {
        let offerings = try await Purchases.shared.offerings()
        return offerings.current?.availablePackages.compactMap { package in
            switch package.identifier {
            case ProductStarter.companion.ID:
                return ProductStarter(
                    id: package.storeProduct.productIdentifier,
                    title: package.storeProduct.localizedTitle,
                    description: package.storeProduct.localizedDescription,
                    price: package.storeProduct.localizedPriceString
                )
                
            case ProductAllIn.companion.ID:
                return ProductAllIn(
                    id: package.storeProduct.productIdentifier,
                    title: package.storeProduct.localizedTitle,
                    description: package.storeProduct.localizedDescription,
                    price: package.storeProduct.localizedPriceString
                )
                
            default:
                let period: ProductPeriod
                if let subscriptionPeriod = package.storeProduct.subscriptionPeriod {
                    switch subscriptionPeriod.unit {
                    case .day:
                        period = ProductPeriodDuration(value: Int32(subscriptionPeriod.value), unit: .day)
                    case .week:
                        period = ProductPeriodDuration(value: Int32(subscriptionPeriod.value), unit: .week)
                    case .month:
                        period = ProductPeriodDuration(value: Int32(subscriptionPeriod.value), unit: .month)
                    case .year:
                        period = ProductPeriodDuration(value: Int32(subscriptionPeriod.value), unit: .year)
                    @unknown default:
                        period = ProductPeriodDuration(value: Int32(subscriptionPeriod.value), unit: .unknown)
                    }
                } else {
                    period = ProductPeriodLifetime()
                }
                
                return ProductOther(
                    id: package.storeProduct.productIdentifier,
                    packageId: package.identifier,
                    title: package.storeProduct.localizedTitle,
                    description: package.storeProduct.localizedDescription,
                    price: package.storeProduct.localizedPriceString,
                    period: period
                )
            }
        } ?? []
    }
    
    func purchase(packageId: String) async throws {
        
        let offerings = try await Purchases.shared.offerings()
        let package = offerings.current?.availablePackages.first(where: { $0.identifier == packageId })
        if(package != nil){
            let result = try await Purchases.shared.purchase(package: package!)
            print("Purchase result: \(result)")
            if result.transaction == nil {
                throw NSError(domain: "Unknown error", code: 1)
            } else if result.userCancelled {
                throw NSError(domain: "User cancelled purchase", code: 2)
            }
            
        }
    }
    
    func restorePurchases() async throws {
        let result = try await Purchases.shared.restorePurchases()
        print("Restore result: \(result)")
    }
}

File: androidApp/src/main/java/com/zenithapps/mobilestack/android/provider/RevenueCatBillingProvider.kt
package com.zenithapps.mobilestack.android.provider

import android.app.Activity
import com.revenuecat.purchases.LogLevel
import com.revenuecat.purchases.PurchaseParams
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.awaitCustomerInfo
import com.revenuecat.purchases.awaitLogIn
import com.revenuecat.purchases.awaitLogOut
import com.revenuecat.purchases.awaitOfferings
import com.revenuecat.purchases.awaitPurchase
import com.revenuecat.purchases.awaitRestore
import com.revenuecat.purchases.models.Period
import com.revenuecat.purchases.models.PurchaseState
import com.zenithapps.mobilestack.android.BuildConfig
import com.zenithapps.mobilestack.model.CustomerBillingInfo
import com.zenithapps.mobilestack.model.Product
import com.zenithapps.mobilestack.provider.BillingProvider

class RevenueCatBillingProvider(
    private val activity: Activity
) : BillingProvider {
    override suspend fun configure(apiKey: String, userId: String?) {
        if (BuildConfig.DEBUG) {
            Purchases.logLevel = LogLevel.DEBUG
        } else {
            Purchases.logLevel = LogLevel.ERROR
        }

        Purchases.configure(
            PurchasesConfiguration.Builder(activity, apiKey)
                .appUserID(userId)
                .build()
        )
    }

    override suspend fun logIn(userId: String, email: String?) {
        Purchases.sharedInstance.awaitLogIn(userId)
        Purchases.sharedInstance.setEmail(email)
    }

    override suspend fun setEmail(email: String) {
        Purchases.sharedInstance.setEmail(email)
    }

    override suspend fun logOut() {
        if (!Purchases.sharedInstance.isAnonymous) {
            Purchases.sharedInstance.awaitLogOut()
        }
    }

    override suspend fun getCustomerBillingInfo(): CustomerBillingInfo {
        val customerInfo = Purchases.sharedInstance.awaitCustomerInfo()
        return CustomerBillingInfo(
            entitlements = customerInfo.entitlements.active.keys.toList(),
            purchases = customerInfo.allPurchasedProductIds.toList(),
            managementUrl = customerInfo.managementURL.toString()
        )
    }

    override suspend fun getProducts(): List<Product> {
        val offerings = Purchases.sharedInstance.awaitOfferings()
        return offerings.current?.availablePackages?.map {
            val periodValue = it.product.period?.value
            val periodUnit = it.product.period?.unit
            Product(
                id = it.product.id,
                packageId = it.identifier,
                title = it.product.name,
                description = it.product.description,
                price = it.product.price.formatted,
                period = if (periodValue != null && periodValue != 0 && periodUnit != null) {
                    Product.Period.Duration(
                        periodValue,
                        when (periodUnit) {
                            Period.Unit.DAY -> Product.PeriodUnit.DAY
                            Period.Unit.WEEK -> Product.PeriodUnit.WEEK
                            Period.Unit.MONTH -> Product.PeriodUnit.MONTH
                            Period.Unit.YEAR -> Product.PeriodUnit.YEAR
                            else -> Product.PeriodUnit.UNKNOWN
                        }
                    )
                } else {
                    Product.Period.Lifetime
                }
            )
        } ?: emptyList()
    }

    override suspend fun purchase(packageId: String) {
        val offerings = Purchases.sharedInstance.awaitOfferings()
        val packageToPurchase =
            offerings.current?.availablePackages?.find { it.identifier == packageId }
                ?: throw Exception("Product not found")
        val result = Purchases.sharedInstance.awaitPurchase(
            purchaseParams = PurchaseParams.Builder(activity, packageToPurchase)
                .build()
        )
        if (result.customerInfo.entitlements.active.isEmpty()) {
            throw Exception("Purchase failed")
        }
        if (result.storeTransaction.purchaseState != PurchaseState.PURCHASED) {
            throw Exception("Purchase pending")
        }
    }

    override suspend fun restorePurchases() {
        Purchases.sharedInstance.awaitRestore()
    }
}